File: .llm/runs/claude-harness-profile-rfc-benchmark-shzhgv--polyglot-protocol-rfc/bench/spikes/k5/task-cancel.go
// K5 Go task — blocking MINSTD compute on main goroutine, stdin control reader goroutine.
// Cooperative cancel at chunk granularity; cancel-ack frame on stop.
package main

import (
	"bufio"
	"encoding/json"
	"os"
	"sync/atomic"
)

const SENTINEL = "\x00NSF\x00"

func frame(obj map[string]any) {
	b, _ := json.Marshal(obj)
	os.Stdout.WriteString(SENTINEL + string(b) + "\n")
}

func main() {
	var cancel atomic.Bool
	go func() {
		sc := bufio.NewScanner(os.Stdin)
		for sc.Scan() {
			var f map[string]any
			if json.Unmarshal(sc.Bytes(), &f) == nil {
				if t, _ := f["t"].(string); t == "cancel" {
					cancel.Store(true)
					return
				}
			}
		}
	}()
	frame(map[string]any{"v": 1, "t": "started"})

	state, acc := uint64(42), uint64(0)
	const TOTAL = uint64(200_000_000_000)
	const CHUNK = uint64(1_000_000)
	for i := uint64(0); i < TOTAL; i += CHUNK {
		for j := uint64(0); j < CHUNK; j++ {
			state = state * 48271 % 2147483647
			acc = (acc + state) % 1000000007
		}
		if cancel.Load() {
			frame(map[string]any{"v": 1, "t": "result", "outcome": "cancelled", "iterations": i})
			return
		}
	}
	frame(map[string]any{"v": 1, "t": "result", "outcome": "ok", "acc": acc})
}

File: .llm/runs/claude-harness-profile-rfc-benchmark-shzhgv--polyglot-protocol-rfc/bench/spikes/k5/task_cancel.py
# K5 python3 task — blocking MINSTD compute on the main thread, stdin control reader on a
# side thread. On CANCEL frame: cooperative stop at chunk granularity, emit cancel-ack frame.
import json
import os
import sys
import threading
import time

SENTINEL = b"\x00NSF\x00"
fd = sys.stdout.fileno()
cancel = threading.Event()


def frame(obj) -> None:
    os.write(fd, SENTINEL + json.dumps(obj, separators=(",", ":")).encode() + b"\n")


def control() -> None:
    for line in sys.stdin:
        try:
            f = json.loads(line)
        except json.JSONDecodeError:
            continue
        if f.get("t") == "cancel":
            cancel.set()
            return


threading.Thread(target=control, daemon=True).start()
frame({"v": 1, "t": "started"})

state, acc = 42, 0
i = 0
TOTAL = 200_000_000  # far longer than the test window; cancel interrupts it
CHUNK = 100_000
while i < TOTAL:
    for _ in range(CHUNK):
        state = state * 48271 % 2147483647
        acc = (acc + state) % 1000000007
    i += CHUNK
    if cancel.is_set():
        frame({"v": 1, "t": "result", "outcome": "cancelled", "ackNs": time.perf_counter_ns(), "iterations": i})
        sys.exit(0)
frame({"v": 1, "t": "result", "outcome": "ok", "acc": acc})
